administrator/components/com_workflow/Field/WorkflowStateField.php

```diff
@@ -0,0 +1,80 @@
+<?php
+/**
+ * @package     Joomla.Administrator
+ * @subpackage  com_workflow
+ *
+ * @copyright   Copyright (C) 2005 - 2017 Open Source Matters, Inc. All rights reserved.
+ * @license     GNU General Public License version 2 or later; see LICENSE.txt
+ */
+namespace Joomla\Component\Workflow\Administrator\Field;
+
+defined('JPATH_BASE') or die;
+
+use Joomla\CMS\Factory;
+use Joomla\CMS\Form\FormHelper;
+
+FormHelper::loadFieldClass('groupedlist');
+
+/**
+ * Workflow States field.
+ *
+ * @since  4.0
+ */
+class WorkflowStateField extends \JFormFieldGroupedList
+{
+	/**
+	 * The form field type.
+	 *
+	 * @var     string
+	 * @since   4.0.0
+	 */
+	protected $type = 'WorkflowState';
+
+	/**
+	 * Method to get the field option groups.
+	 *
+	 * @return  array  The field option objects as a nested array in groups.
+	 *
+	 * @since   4.0.0
+	 * @throws  \UnexpectedValueException
+	 */
+	protected function getGroups()
+	{
+		$db    = Factory::getDbo();
+		$query = $db->getQuery(true);
+
+		// Select distinct states for existing articles
+		$query
+			->select('DISTINCT ' . $db->qn('ws.id', 'workflow_state_id'))
```

Comment:
Perhaps use ```->group($db->qn('ws.id'))``` instead of ```DISTINCT``` so you can do the 4 selects into one:

`
->select($db->qn(['ws.id', 'ws.title', 'w.title', 'w.id'], ['workflow_state_id', 'workflow_state_title', 'workflow_title', 'workflow_id']))
`

Reply:
using group didn't work